sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/LogicalWriteInfo.java

```diff
@@ -45,4 +45,18 @@ public interface LogicalWriteInfo {
    * the schema of the input data from Spark to data source.
    */
   StructType schema();
+
+  /**
+   * the schema of the input metadata from Spark to data source.
+   */
+  default StructType metadataSchema() {
+    return null;
```

Comment:
The default implementation is purely for compatibility.

Reply:
Nit: I usually really want to avoid returning `null`.  Even throw an exception is better than null. Probably this is only my opinion.

Reply:
I am open to discuss what value should indicate no row ID or metadata schema.

Reply:
I see `Optional` in a few other places in the connector API. I could switch to that.

Reply:
Let me see how the implementation use this API if there is any.

Reply:
Returning null is pretty risky as you don't know when/where a NPE will happen. I'd prefer throw `UnsupportedOperation` by default.

Reply:
@amaliujia @cloud-fan, we can do something like this.

```
/**
 * the schema of the ID columns from Spark to data source.
 */
default Optional<StructType> rowIdSchema() {
  throw new UnsupportedOperationException(
      getClass().getName() + " does not implement rowIdSchema");
}

/**
 * the schema of the input metadata from Spark to data source.
 */
default Optional<StructType> metadataSchema() {
  throw new UnsupportedOperationException(
      getClass().getName() + " does not implement metadataSchema");
}
```

Now the question is what to report in `schema()` for delta-based DELETE operations, where we do not pass the row, we just pass row ID and metadata. One option is to report an empty struct but let me know if you have other ideas.

The way I was approaching it initially:

```
schema() -> the row schema for new records (only MERGE adds new records)
rowIdSchema() -> the schema for row ID passed to data sources to mark a record as deleted/updated
metadataSchema() -> the schema of projected metadata columns that contain some extra info about the row that is being deleted/updated
```

Reply:
Please proceed to update this PR according to your proposed way, @aokolnychyi .

Reply:
I've updated this PR and the reference implementation in PR #38005.